### ISSUE TITLE
Single label queries sent too often

### ISSUE THREAD

**Issue by pemensik** [systemd/systemd#23713]
**systemd version the issue has been seen with**
 > systemd-resolved-250.7-1.fc36.x86_64

**Used distribution**
 > Fedora 36

**Linux kernel version used** (`uname -a`)

 > `Linux menryzen 5.17.12-300.fc36.x86_64 #1 SMP PREEMPT Mon May 30 16:56:53 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > Fast response to 

**Unexpected behaviour you saw**
 > I did ``getent ahosts -s resolve nx-host2`` and recorded, what queries it will send

**Steps to reproduce the problem**
 > Record outgoing queries and request single label name. Watch what queries are made.
- ``tcpdump -nv port domain or port llmnr``
- ``getent ahosts -s resolve nx-host3``

**Additional program output to the terminal or log subsystem illustrating the issue**

```text
tcpdump: listening on virbr0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
17:34:28.184892 IP (tos 0x0, ttl 64, id 35177, offset 0, flags [none], proto UDP (17), length 75)
    192.168.122.26.32842 > 192.168.122.1.domain: 58200+ [1au] A? nx-host3.home.arpa. (47)
17:34:28.184987 IP (tos 0x0, ttl 64, id 27800, offset 0, flags [none], proto UDP (17), length 75)
    192.168.122.26.47216 > 192.168.122.1.domain: 10599+ [1au] AAAA? nx-host3.home.arpa. (47)
17:34:28.192166 IP (tos 0x0, ttl 64, id 6573, offset 0, flags [DF], proto UDP (17), length 75)
    192.168.122.1.domain > 192.168.122.26.32842: 58200 NXDomain 0/0/1 (47)
17:34:28.192592 IP (tos 0x0, ttl 64, id 35178, offset 0, flags [none], proto UDP (17), length 64)
    192.168.122.26.32842 > 192.168.122.1.domain: 58200+ A? nx-host3.home.arpa. (36)
17:34:28.194650 IP (tos 0x0, ttl 64, id 6574, offset 0, flags [DF], proto UDP (17), length 75)
    192.168.122.1.domain > 192.168.122.26.47216: 10599 NXDomain 0/0/1 (47)
17:34:28.194969 IP (tos 0x0, ttl 64, id 27801, offset 0, flags [none], proto UDP (17), length 64)
    192.168.122.26.47216 > 192.168.122.1.domain: 10599+ AAAA? nx-host3.home.arpa. (36)
17:34:28.197265 IP (tos 0x0, ttl 64, id 6575, offset 0, flags [DF], proto UDP (17), length 64)
    192.168.122.1.domain > 192.168.122.26.32842: 58200 NXDomain 0/0/0 (36)
17:34:28.199762 IP (tos 0x0, ttl 64, id 6578, offset 0, flags [DF], proto UDP (17), length 64)
    192.168.122.1.domain > 192.168.122.26.47216: 10599 NXDomain 0/0/0 (36)
17:34:28.305416 IP6 (flowlabel 0x47df0, hlim 255, next-header UDP (17) payload length: 34) fe80::7e25:31d9:61a5:b92e.hostmon > ff02::1:3.hostmon: [bad udp cksum 0xc88d -> 0x0890!] UDP, length 26
17:34:28.305526 IP (tos 0x0, ttl 255, id 18814, offset 0, flags [none], proto UDP (17), length 54)
    192.168.122.26.hostmon > 224.0.0.252.hostmon: UDP, length 26
17:34:29.382936 IP6 (flowlabel 0x47df0, hlim 255, next-header UDP (17) payload length: 34) fe80::7e25:31d9:61a5:b92e.hostmon > ff02::1:3.hostmon: [bad udp cksum 0xc88d -> 0x0890!] UDP, length 26
17:34:29.382953 IP (tos 0x0, ttl 255, id 18938, offset 0, flags [none], proto UDP (17), length 54)
    192.168.122.26.hostmon > 224.0.0.252.hostmon: UDP, length 26
17:34:30.632602 IP6 (flowlabel 0x47df0, hlim 255, next-header UDP (17) payload length: 34) fe80::7e25:31d9:61a5:b92e.hostmon > ff02::1:3.hostmon: [bad udp cksum 0xc88d -> 0x0890!] UDP, length 26
17:34:30.632625 IP (tos 0x0, ttl 255, id 19329, offset 0, flags [none], proto UDP (17), length 54)
    192.168.122.26.hostmon > 224.0.0.252.hostmon: UDP, length 26
```
It shows too quick repetition of original query, it sends another request without waiting enough time for original response. It waits very short time before retry. Also it queries again the same name it got reply already. NXDOMAIN is a valid reply. It is discutable whether it makes sense to query other servers. But it does not make sense to query the same server again for the same name. Reply should never change so fast. Good cache will cache also negative replies for some short time. It does not seem resolved cache even for 2 seconds. That should be fixed.

I think it should wait at least for 250ms before first retry. Unless upstream resolver is authoritative for that zone, it is unlikely it will respond fast enough. It should not generate more queries than without the cache present. I think cache presence should decrease number of queries sent to upstream.

```
$ resolvectl
Global
       Protocols: LLMNR=resolve -mDNS -DNSOverTLS DNSSEC=no/unsupported
resolv.conf mode: stub

Link 2 (enp1s0)
    Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
         Protocols: +DefaultRoute +LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
Current DNS Server: 192.168.122.1
       DNS Servers: 192.168.122.1
        DNS Domain: home.arpa
```

Low severity.

**Comment by poettering** [comment]
so the 6 packets at the end are mdns, not classic dns. tcpdump shows that as "hostmon", because /etc/services is such garbage.

At the top you see an A and an AAAA query plus responses. And then you see them a second time but with much shorter datagram length (75 → 64). My educated guess: this was a downgrade of some sorts, server didn't support what resolved tried first.

Without proper debug logs we cannot verify that though. If we had a proper pcap we might be able to verify this too. 

Generally: please always provide debug logs, otherwise things like this are not actionable.

Can you provide debug logs or pcap? Otherwise I think we should just close this, I see nothing we could do here.

**Comment by pemensik** [comment]
I have provided exact steps to reproduce the reported issue. Unless it does not happen on your network, then you can make pcap yourself quite easily. Have you tried that with reported version? Have you received significantly different results?

It has turned out the negative answer re-query was issue reported separately #26967 and fix for that is already merged in #27078.

hostmon is quite strange, especially when only llmnr is registered at [iana.org](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?search=5355). Filled bug https://bugzilla.redhat.com/show_bug.cgi?id=2216914 on that.

**Comment by pemensik** [comment]
This were actionable just fine, but anyway it turns out it is duplicate of https://github.com/systemd/systemd/issues/26967, which already were fixed. Closing this one
